modules/apps/portal/portal-servlet-test/src/testIntegration/java/com/liferay/portal/servlet/filters/multihost/locale/test/MultiHostLocaleFilterTest.java

```diff
@@ -0,0 +1,263 @@
+			_virtualHostLocalService.getVirtualHost(hostname).getLanguageId());
+
+		Assert.assertEquals(((LayoutSet)
+			_mockHttpServletRequest.getAttribute(
+			_VIRTUAL_HOST_LAYOUT_SET)).getLayoutSetId(), _layoutSetId);
+	}
+
+	@Test
+	public void testMultiHostName1() throws Exception{
+
+		_virtualHost = "testhost.fr,testhost.jp";
+
+		_treeMap = toTreeMap(_virtualHost);
+		_layoutSetLocalService.updateVirtualHosts(
+			_groupId, false, _treeMap);
+
+		for (String host : _treeMap.keySet()) {
+			_setupRequest(host, _treeMap.get(host));
+
+			_invokeRequestFilter();
+
+			Assert.assertEquals(((LayoutSet)
+				_mockHttpServletRequest.getAttribute(
+				_VIRTUAL_HOST_LAYOUT_SET)).getLayoutSetId(), _layoutSetId);
+
+			_cleanUpRequest();
+		}
+	}
+
+	@Test
+	public void testMultiHostName2() throws Exception{
+
+		_virtualHost = "testhost.de";
+
+		_treeMap = toTreeMap(_virtualHost);
+		_layoutSetLocalService.updateVirtualHosts(
+			_groupId, false, _treeMap);
+
+		String hostname = _treeMap.firstKey();
+		_setupRequest(hostname, _treeMap.get(hostname));
+
+		_invokeRequestFilter();
+
+		Assert.assertEquals(((LayoutSet)
+			_mockHttpServletRequest.getAttribute(
+			_VIRTUAL_HOST_LAYOUT_SET)).getLayoutSetId(), _layoutSetId);
+	}
+
+	private void _invokeRequestFilter() throws Exception {
+		_absoluteRedirectsFilter.doFilterTry(
+			_mockHttpServletRequest, _mockHttpServletResponse);
+	}
+
+	private void _setupRequest(String hostname, String languageId)
+		throws AvailableLocaleException {
+
+		if (Validator.isNotNull(languageId)) {
+			List<Locale> locales = new ArrayList<>();
+
+			Locale locale = LocaleUtil.fromLanguageId(languageId);
+
+			locales.add(locale);
+
+			if (!_availableLocales.contains(locale)) {
+				throw new AvailableLocaleException(languageId);
+			}
+
+			_mockHttpServletRequest.setPreferredLocales(locales);
+		}
+
+		_mockHttpServletRequest.setRemoteHost(hostname);
+		_mockHttpServletRequest.setServerName(hostname);
+		_mockHttpServletRequest.setServerPort(_SERVER_PORT);
+		_mockHttpServletRequest.setRequestURI(_WEB_GUEST);
+		_mockHttpServletRequest.addHeader(_HOST, hostname);
+	}
+
+	private void _cleanUpRequest() {
+		_mockHttpServletRequest.invalidate();
+		_mockHttpServletRequest = new MockHttpServletRequest();
+	}
+
+	private TreeMap<String, String> toTreeMap(String virtualHost)
+		throws AvailableLocaleException {
+
+		TreeMap<String, String> treeMap = new TreeMap<>();
+
+		for (String part : StringUtil.split(virtualHost)) {
+			String[] subparts = StringUtil.split(part, ';');
+
+			if (subparts.length == 2) {
+				String languageId = subparts[1];
+
+				Locale locale = LocaleUtil.fromLanguageId(languageId);
+
+				if (!_availableLocales.contains(locale)) {
+					throw new AvailableLocaleException(languageId);
+				}
+
+				treeMap.put(subparts[0], subparts[1]);
+			}
+			else if (subparts.length == 1) {
+				treeMap.put(subparts[0], StringPool.BLANK);
+			}
+		}
+
+		return treeMap;
+	}
```

Comment:
I think you could simply create a TreeMap directly without having all this logic.